backend/flask/app.py

```diff
@@ -124,20 +185,120 @@ def create_link_token(userId):
         return jsonify(json.loads(e.body)), e.status
 
 
-@app.route('/exchange_public_token', methods=['POST'])
-def exchange_public_token():
-    data = request.get_json()
-    public_token = data.get('public_token')
-    if not public_token:
-        return jsonify({'error': 'Public token is required'}), 400
-
+@app.route('/api/exchange_public_token/<userId>', methods=['POST'])
+def exchange_public_token(userId):
+    public_token = request.json['public_token']
     try:
-        exchange_response = plaid_client.item_public_token_exchange(public_token)
+        exchange_request = ItemPublicTokenExchangeRequest(
+            public_token=public_token
+        )
+        exchange_response = plaid_client.item_public_token_exchange(exchange_request)
         access_token = exchange_response['access_token']
-        return jsonify({'access_token': access_token}), 200
-    except plaid.ApiException as e:
-        return jsonify(json.loads(e.body)), e.status 
+        item_id = exchange_response['item_id']
+        user = User.query.get(userId)
+        # store the user's access token in user table, an access token provides repeated access to user's plaid-data
+        user.token = access_token
+        db.session.commit()
+        
+        return jsonify({
+            'access_token': access_token,
+            'item_id': item_id,
+            'public_token_exchange': 'complete'
+        })
+    except Exception as e:
+        return jsonify({'error': str(e)}), 500
+
+@app.route('/api/transactions/sync/<userId>', methods=['POST'])
+def transactions_sync(userId):
+    try:
+        # get the user's access_token from the user table
+        user = User.query.get(userId)
+        request = TransactionsSyncRequest(
+            access_token=user.token
+        )
+        transactions = plaid_client.transactions_sync(request)
+        # transaction object is by default non-serialable hence why we transform it to a dictionary
+        transactions_data = {
+            "transactions": [transaction.to_dict() for transaction in transactions.added]
+        }
+        clean_data =  clean_transaction_data(transactions_data)
+        user_transaction_data = aggregate_user_data(clean_data)
+        db_status = save_transaction(userId, user_transaction_data)
+
+        return jsonify({'message': db_status, 'data': user_transaction_data})
+    except Exception as e:
+        return jsonify({'error': str(e)}), 500
+
+
+@app.route('/transactions/<userId>', methods=['GET'])
+def get_latest_transaction(userId):
+    user = User.query.get(userId)
+    transaction = user.transactions[-1]
```

Comment:
This would likely lead to performance issues if a user has been using my site for a long time and has a lot of rows to read from but I wasn't able to think of a better solution to fetch the "newest" row. I'm sure there's a way to get one using more SQL-specific methods I'm not aware of it. This is maybe a "good enough for now" thing that a tab should be kept on to find a more optimal solution to.

Reply:
Does each row have a "date" column? Could you add an index on that column and and use `ORDER BY` in your query?

Reply:
Ok yeah that's a good solution – since I'm assuming SQL uses a binary tree or something like that to store data so looking up like that would be logarithmic. I'll add that change onto my list. Thanks!